File: sqrtx.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x in (0,1):
            return x
        pre, post = 1, x
        while True:
            if ((pre+post)//2)**2 <= x and ((pre+post)//2+1)**2 > x:
                return (pre+post)//2
            elif ((pre+post)//2)**2 > x:
                post = (pre+post)//2
                #last = post
            elif ((pre+post)//2)**2 < x:
                pre = (pre+post)//2
                #last = pre
            

File: cpp/sqrtx.cpp
class Solution {
public:
    int mySqrt(int x) {
        if (x == 0 or x == 1) return x;
        int i =  1, j = x;
        while (1) {
            int mid = i + (j-i)/2;
            if (pow(mid, 2) <= x and pow(mid+1, 2) > x) return mid;
            else if (pow(mid, 2) < x) {
                i = mid + 1;
            }
            else {
                j = mid - 1;
            }
        }
    }
};
